main/guides/zoe/contract-upgrade.md

```diff
@@ -0,0 +1,168 @@
+
+:::
+
+<<< @/snippets/zoe/src/02b-state-durable.js#provide-cell
+
+::: details What happens if we don't use baggage?
+
+When the contract instance is restarted, it gets a fresh [heap](../js-programming/#vats-the-unit-of-synchrony), so [ordinary heap state](./contract-basics.html#state) does not survive upgrade. This implementation does not persist the effect of `E(publicFacet).set(2)` from the first incarnation:
+
+<<< @/snippets/zoe/src/02-state.js#heap-state{2}
+
+Also, it creates a fresh `publicFacet` object each time. So messages
+from clients that try to use the `publicFacet` from the first incarnation
+will fail to reach the `publicFacet` created in later incarnations.
+
+<<< @/snippets/zoe/src/02-state.js#fresh-export{2}
+
+:::
+
+### Kinds
+
+Use `zone.exoClass()` to define state and methods of kinds of durable objects such as `ValueCell`:
+
+::: details import { makeDurableZone } ...
+
+<<< @/snippets/zoe/src/02b-state-durable.js#import-provide-zone{1}
+
+:::
+
+<<< @/snippets/zoe/src/02b-state-durable.js#exo
+
+Now we have all the parts of an upgradable contract.
+
+::: details full contract listing
+
+<<< @/snippets/zoe/src/02b-state-durable.js#contract
+
+:::
+
+We can then upgrade it to have another method:
+
+```js
+const CellI = M.interface('ValueCell', {
+  ...
+  incr: M.call(M.number()).returns(),
+});
+
+  const makeCell = zone.exoClass('ValueCell', CellI, () => ({ value: 0 }), {
+    ...
+    incr(delta) {
+      this.state.value += delta;
+    },
+  });
+```
+
+::: tip Notes
+
+- Once the state is defined by the `init` function (3rd arg), properties cannot be added or removed.
+- Values of state properties must be serializable.
+- Values of state properties are hardened on assignment.
+- You can replace the value of a state property (e.g. `state.zot = [...state.zot, 'last']`), but you cannot mutate it (`state.zot.push('last')`).
+- The tag (1st arg) is used to form a key in `baggage`, so take care to avoid collisions. `zone.subZone()` may be used to partition namespaces.
+- See also [defineExoClass](https://endojs.github.io/endo/functions/_endo_exo.defineExoClass.html) for further detail `zone.exoClass`.
+- To define multiple objects that share state, use `zone.exoClassKit`.
+  - See also [defineExoClassKit](https://endojs.github.io/endo/functions/_endo_exo.defineExoClassKit.html)
+- For an extended test / example, see [test-coveredCall-service-upgrade.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/test/swingsetTests/upgradeCoveredCall/test-coveredCall-service-upgrade.js).
+
+:::
+
+## Crank
+
+Define all exo classes/kits before any incoming method calls from other vats -- in the first "crank".
+
+::: tip Note
+
+- For more on crank constraints, see [Virtual and Durable Objects](https://github.com/Agoric/agoric-sdk/blob/master/packages/SwingSet/docs/virtual-objects.md#virtual-and-durable-objects) in [SwingSet docs](https://github.com/Agoric/agoric-sdk/tree/master/packages/SwingSet/docs)
+
+:::
```

Comment:
This is great, I learned a few things! 

Two suggestions maybe worth implementing - 

1. Do we want to discuss `durability` vs `upgradability` (if not slated for a different section)?. Towards the beginning it may be worthwhile to mention why you may want to make parts of your contract durable (but not necessarily upgradable)
2. May be worth adding upgradability tips/best practices - e.g. backwards compatibility, optional extra arguments, making sure the prepare script doesn't reset counts / state (unless intended).

Reply:
> why you may want to make parts of your contract durable (but not necessarily upgradable)

Is there any reason besides upgrade for durability? I can't think of any.

> backwards compatibility, optional extra arguments

Got any specific suggestions? I'm struggling to find a concise statement of best practices. The semver project punted on a definition, for example:

 - https://github.com/semver/semver/issues/256

https://www.linkedin.com/advice/0/what-best-practices-ensuring-backward-compatibility-i44we has the most likely looking title, but the content doesn't look that good

> making sure the prepare script doesn't reset counts / state

that seems to fall out naturally from the zone/exo API. It doesn't seem useful to say "don't go out of your way to shoot yourself in the foot."

Reply:
> Is there any reason besides upgrade for durability? I can't think of any.

I've been under the impression "making my contract survive a chain upgrade" and "make my contract upgradable (new version, functionality changes)" are two separate things, but seem to be mistaken

Reply:
Contract survive chain upgrades without being durable, but it _is_ useful to be able to restart a contract to clear out the heap. That's noted in the **re-using the same bundle** details box.